src/modules/power_profiles_daemon.cpp

```diff
@@ -0,0 +1,214 @@
+  Gio::DBus::Proxy::create_for_bus(Gio::DBus::BusType::BUS_TYPE_SYSTEM, "net.hadess.PowerProfiles",
+                                   "/net/hadess/PowerProfiles", "net.hadess.PowerProfiles",
+                                   sigc::mem_fun(*this, &PowerProfilesDaemon::busConnectedCb));
+}
+
+PowerProfilesDaemon::~PowerProfilesDaemon() {
+  if (powerProfileChangeSignal_.connected()) {
+    powerProfileChangeSignal_.disconnect();
+  }
+  if (powerProfilesProxy_) {
+    powerProfilesProxy_.reset();
+  }
+}
+
+void PowerProfilesDaemon::busConnectedCb(Glib::RefPtr<Gio::AsyncResult>& r) {
+  try {
+    powerProfilesProxy_ = Gio::DBus::Proxy::create_for_bus_finish(r);
+    using GetAllProfilesVar = Glib::Variant<std::tuple<Glib::ustring>>;
+    auto callArgs = GetAllProfilesVar::create(std::make_tuple("net.hadess.PowerProfiles"));
+
+    auto container = Glib::VariantBase::cast_dynamic<Glib::VariantContainerBase>(callArgs);
+    powerProfilesProxy_->call("org.freedesktop.DBus.Properties.GetAll",
+                              sigc::mem_fun(*this, &PowerProfilesDaemon::getAllPropsCb), container);
+    // Connect active profile callback
+  } catch (const std::exception& e) {
+    spdlog::error("Failed to create the power profiles daemon DBus proxy");
+  }
+}
+
+// Callback for the GetAll call.
+//
+// We're abusing this call to make sure power-profiles-daemon is
+// available on the host. We're not really using
+void PowerProfilesDaemon::getAllPropsCb(Glib::RefPtr<Gio::AsyncResult>& r) {
+  try {
+    auto _ = powerProfilesProxy_->call_finish(r);
+    // Power-profiles-daemon responded something, we can assume it's
+    // available, we can safely attach the activeProfile monitoring
+    // now.
+    connected_ = true;
+    powerProfileChangeSignal_ = powerProfilesProxy_->signal_properties_changed().connect(
+        sigc::mem_fun(*this, &PowerProfilesDaemon::profileChangedCb));
+    populateInitState();
+    dp.emit();
+  } catch (const std::exception& err) {
+    spdlog::error("Failed to query power-profiles-daemon via dbus: {}", err.what());
+  }
+}
+
+void PowerProfilesDaemon::populateInitState() {
+  // Retrieve current active profile
+  Glib::Variant<std::string> profileStr;
+  powerProfilesProxy_->get_cached_property(profileStr, "ActiveProfile");
+
+  // Retrieve profiles list, it's aa{sv}.
+  using ProfilesType = std::vector<std::map<Glib::ustring, Glib::Variant<std::string>>>;
+  Glib::Variant<ProfilesType> profilesVariant;
+  powerProfilesProxy_->get_cached_property(profilesVariant, "Profiles");
+  Glib::ustring name;
+  Glib::ustring driver;
+  Profile profile;
+  for (auto& variantDict : profilesVariant.get()) {
+    if (auto p = variantDict.find("Profile"); p != variantDict.end()) {
+      name = p->second.get();
+    }
+    if (auto d = variantDict.find("Driver"); d != variantDict.end()) {
+      driver = d->second.get();
+    }
+    profile = {name, driver};
+    availableProfiles_.push_back(profile);
```

Comment:
`availableProfiles_.emplace_back(name, driver);` would avoid one temporary object.

And if you move name/driver declaration into the loop (which you might want to do anyways to avoid accidentally reusing values from previous iterations), it'll be safe to write `if (!name.empty()) { availableProfiles_.emplace_back(std::move(name), std::move(driver)); }`